guests/python/src/lib.rs

```diff
@@ -3,82 +3,227 @@
         let ScalarFunctionArgs {
             args,
-            arg_fields: _,
-            number_rows: _,
-            return_field: _,
+            arg_fields,
+            number_rows,
+            return_field,
         } = args;
 
-        if !args.is_empty() {
-            return exec_err!("test expects no arguments");
+        let return_dt = self.python_function.signature.return_type.t.data_type();
+        if return_field.data_type() != &return_dt {
+            return exec_err!(
+                "`{}` returns {} but was asked to produce {}",
+                self.name(),
+                return_dt,
+                return_field.data_type()
+            );
+        }
+        if !return_field.is_nullable() {
+            return exec_err!(
+                "`{}` returns nullable data but was asked not to do so",
+                self.name()
+            );
+        }
+
+        if arg_fields.len() != self.python_function.signature.parameters.len() {
+            return exec_err!(
+                "`{}` expects {} parameters (passed as fields) but got {}",
+                self.name(),
+                self.python_function.signature.parameters.len(),
+                arg_fields.len()
+            );
+        }
+        for (i, (arg_field, param)) in arg_fields
+            .iter()
+            .zip(&self.python_function.signature.parameters)
+            .enumerate()
+        {
+            let param_dt = param.t.data_type();
+            if arg_field.data_type() != &param_dt {
+                return exec_err!(
+                    "argument field {} of `{}` should be {} but got {}",
+                    i + 1,
+                    self.name(),
+                    param_dt,
+                    arg_field.data_type()
+                );
+            }
+        }
+
+        if args.len() != self.python_function.signature.parameters.len() {
+            return exec_err!(
+                "`{}` expects {} parameters (passed as args) but got {}",
+                self.name(),
+                self.python_function.signature.parameters.len(),
+                args.len()
+            );
         }
 
-        let s = Python::with_gil(|py| {
-            let sys = py.import("sys")?;
-            let version: String = sys.getattr("version")?.extract()?;
+        let arrays = args
+            .into_iter()
+            .map(|column_value| column_value.to_array(number_rows))
+            .collect::<Result<Vec<_>, _>>()?;
 
-            let locals = [("os", py.import("os")?)].into_py_dict(py)?;
-            let code = c_str!("os.getenv('USER') or os.getenv('USERNAME') or 'Unknown'");
-            let user: String = py.eval(code, None, Some(&locals))?.extract()?;
+        Python::with_gil(|py| {
+            let mut parameter_iters = arrays
+                .iter()
+                .zip(&self.python_function.signature.parameters)
+                .map(|(array, t)| t.arrow_to_python(array, py))
```

Comment:
Does `arrow_to_python` need to be called within the gil? If so, could you add a comment on the method? If not, could you move this outside the gil lock?